pkg/storage/storagewrappers/sharediterator/shared_iterator_datastore.go

```diff
@@ -196,74 +209,60 @@ func (sf *IteratorDatastore) ReadStartingWithUser(
+		sharedIteratorBypassed.WithLabelValues(storagewrappersutil.OperationReadStartingWithUser).Inc()
+		return sf.RelationshipTupleReader.ReadStartingWithUser(ctx, store, filter, options)
+	}
 
-	if found {
-		keyItem.counter++
-		sf.internalStorage.mu.Unlock()
-		keyItem.iter.mu.Lock()
+	newStorageItem := new(storageItem)
 
-		span.SetAttributes(attribute.Bool("found", true))
+	timer := time.AfterFunc(sf.maxAdmissionTime, func() {
+		if sf.internalStorage.iters.CompareAndDelete(cacheKey, newStorageItem) {
+			sf.internalStorage.ctr.Add(-1)
+		}
+	})
 
-		// by the time we have access to keyItem.iter.mu, we know that
-		// the inner is already set OR the initializationErr is set (for which clone
-		// will fail).
-		item, err := keyItem.iter.clone()
+	newStorageItem.producer = func() (*sharedIterator, error) {
+		it, err := sf.RelationshipTupleReader.ReadStartingWithUser(ctx, store, filter, options)
 		if err != nil {
-			// This needs to be unlocked before sf.RelationshipTupleReader.ReadStartingWithUser
-			// to reduce the time holding lock across expensive operation
-			keyItem.iter.mu.Unlock()
-			sf.deref(cacheKey)
-			telemetry.TraceError(span, err)
-			return sf.RelationshipTupleReader.ReadStartingWithUser(ctx, store, filter, options)
+			return nil, err
 		}
-		keyItem.iter.mu.Unlock()
+
+		newIterator := newSharedIterator(it, func() {
+			timer.Stop()
+			if sf.internalStorage.iters.CompareAndDelete(cacheKey, newStorageItem) {
+				sf.internalStorage.ctr.Add(-1)
+			}
+		})
+
+		sharedIteratorCount.Inc()
+
+		span.SetAttributes(attribute.Bool("found", false))
+
 		sharedIteratorQueryHistogram.WithLabelValues(
-			storagewrappersutil.OperationReadStartingWithUser, sf.method, "true",
+			storagewrappersutil.OperationReadStartingWithUser, sf.method, "false",
 		).Observe(float64(time.Since(start).Milliseconds()))
-		return item, nil
-	}
-	if len(sf.internalStorage.iters) >= sf.internalStorage.limit {
-		span.SetAttributes(attribute.Bool("overlimit", true))
 
-		sf.internalStorage.mu.Unlock()
-		sharedIteratorBypassed.WithLabelValues(storagewrappersutil.OperationReadStartingWithUser).Inc()
-		// we cannot share this iterator because we have reached the size limit.
-		return sf.RelationshipTupleReader.ReadStartingWithUser(ctx, store, filter, options)
+		return newIterator, err
 	}
 
-	newIterator := newSharedIterator(sf, cacheKey, sf.watchdogTTL, sf.maxAdmissionTime, sf.iteratorTargetSize)
-	newIterator.mu.Lock()
-
-	sf.internalStorage.iters[cacheKey] = &internalSharedIterator{
-		counter: 1,
-		iter:    newIterator,
+	value, loaded := sf.internalStorage.iters.LoadOrStore(cacheKey, newStorageItem)
+	if !loaded {
+		sf.internalStorage.ctr.Add(1)
 	}
-	sharedIteratorCount.Inc()
-	sf.internalStorage.mu.Unlock()
-	span.SetAttributes(attribute.Bool("found", false))
-
-	actual, err := sf.RelationshipTupleReader.ReadStartingWithUser(ctx, store, filter, options)
+	item, _ := value.(*storageItem)
+	it, err := item.unwrap()
```

Comment:
Fancy. It basically boots itself conditionally by using your LoadOrStore + unwrap. If the first instance it will just provision the iterator, if not it would clone it. We gotta document this behavior so its clear.